IVIM Workflow

ShreyasFadnavis opened the pull request:
- [x] Testing
- [x] Adding remaining parameters

pep8speaks commented:
Hello @ShreyasFadnavis, Thank you for submitting the Pull Request !

 - In the file [`dipy/workflows/reconst.py`](https://github.com/nipy/dipy/blob/27258e856a633d75d0065cdbb42a43bf2469f3cb/dipy/workflows/reconst.py), following are the PEP8 issues :

> [Line 965:13](https://github.com/nipy/dipy/blob/27258e856a633d75d0065cdbb42a43bf2469f3cb/dipy/workflows/reconst.py#L965): [E117](https://duckduckgo.com/?q=pep8%20E117) over-indented

Do see the [DIPY coding Style guideline](https://github.com/nipy/dipy/blob/master/doc/devel/coding_style_guideline.rst)

ShreyasFadnavis commented:
@arokem can you take a look and let me know the changes?

skoudoro commented:
Overall, it looks good to me, Thank you @ShreyasFadnavis. Here a couple of comments:

- Can you rebase your PR?
- Can you fix all pep8? Look at [pep8check comments](https://github.com/nipy/dipy/pull/1739#issuecomment-463038039)
- I think, you should remove `small_ivim.nii.gz`, `small_ivim.bval`, `small_ivim.bvec`. It will be better to generate and save it on your temporary folder, run your workflow with that and then delete them.

ShreyasFadnavis commented:
> Overall, it looks good to me, Thank you @ShreyasFadnavis. Here a couple of comments:
> 
> * Can you rebase your PR?
> * Can you fix all pep8? Look at [pep8check comments](https://github.com/nipy/dipy/pull/1739#issuecomment-463038039)
> * I think, you should remove `small_ivim.nii.gz`, `small_ivim.bval`, `small_ivim.bvec`. It will be better to generate and save it on your temporary folder, run your workflow with that and then delete them.

Thank you @skoudoro ! The reason I added `ivim_small` is because IVIM needs a special kind of data, I'm not sure of how I can generate them.

I have fixed the pep8 in the last push!

ShreyasFadnavis commented on dipy/workflows/reconst.py:
```diff
@@ -861,3 +862,117 @@ def get_fitted_tensor(self, data, mask, bval, bvec, b0_threshold=50):
+            if not save_metrics:
+                save_metrics = ['S0_est', 'f_est', 'D_star_est', 'D_est']
+
+            S0_est = ivimfit.model_params[..., 0]
```
@arokem I wanted your opinion on this.. Does this look correct to you?
Any advice would be much appreciated!

arokem replied:
Yes. Looks right. You can also point to [this property](https://github.com/nipy/dipy/blob/master/dipy/reconst/ivim.py#L541) if you'd like to be a bit tidier about it.

arokem commented on dipy/workflows/reconst.py:
```diff
@@ -861,3 +862,117 @@ def get_fitted_tensor(self, data, mask, bval, bvec, b0_threshold=50):
+            S0_est = ivimfit.model_params[..., 0]
+            f_est = ivimfit.model_params[..., 1]
+            D_star_est = ivimfit.model_params[..., 2]
+            D_est = ivimfit.model_params[..., 3]
```
And same for these: you point to [these properties](https://github.com/nipy/dipy/blob/master/dipy/reconst/ivim.py#L544-L554)

I'd also advise using the same nomenclature that we used there (e.g., `S0_predicted`, `perfusion_fraction` and so forth), unless there is some compelling reason to refer these by different names here.

ShreyasFadnavis replied:
Didn't realize, will do! :+1:

arokem commented on dipy/sims/voxel.py:
```diff
@@ -777,7 +777,7 @@ def multi_tensor_odf(odf_verts, mevals, angles, fractions):
 
 def single_tensor_rtop(evals=None, tau=1.0 / (4 * np.pi ** 2)):
-    """Simulate a Multi-Tensor rtop.
+    """Simulate a Single-Tensor rtop.
```
Good catch!

arokem commented on dipy/sims/voxel.py:
```diff
@@ -997,6 +997,44 @@ def multi_tensor_msd(mf, mevals=None, tau=1 / (4 * np.pi ** 2)):
         msd += f * single_tensor_msd(mevals[j], tau=tau)
     return msd
 
+def ivim_signal(gtab, S0, f, D_star, D):
```
This seems to duplicate functionality already implemented [in the ivim module](https://github.com/nipy/dipy/blob/master/dipy/reconst/ivim.py#L21). Could we reuse that instead?

ShreyasFadnavis replied:
Yes, I have replicated this in `sims` so that I can use it to test the MIX module which is another PR. 
I can take it down from this PR I guess! Let me know :)

arokem replied:
Yes. Please remove this and use the implementation in the ivim module as needed.

ShreyasFadnavis replied:
> Yes. Please remove this and use the implementation in the ivim module as needed.

Will do!

arokem commented on dipy/workflows/tests/test_reconst_ivim.py:
```diff
@@ -0,0 +1,88 @@
+        np.savetxt(tmp_bvec_path, bvecs.T)
+
+        ivim_flow._force_overwrite = True
+# TODO
```
Is this still a TODO item?

ShreyasFadnavis replied:
Yes, @skoudoro suggested this to make Travis happy, due to versioning issues of Python.

arokem replied:
So should this code be uncommented?

ShreyasFadnavis replied:
I think I will take it down and add it later, once Python 2.7 is discontinued. @skoudoro,  sounds good?

ShreyasFadnavis commented on dipy/workflows/reconst.py:
```diff
@@ -941,27 +943,30 @@ def run(self, input_files, bvalues_files, bvectors_files, split_b_D=400,
+                save_metrics = ['S0_predicted', 'perfusion_fraction', 'D_star',
+                                'D']
 
-            S0_est = ivimfit.model_params[..., 0]
```
@arokem sorry for the delay, but this is what you wanted me to do right? Let me know if I misunderstood!

arokem commented on dipy/workflows/reconst.py:
```diff
@@ -861,3 +862,122 @@ def get_fitted_tensor(self, data, mask, bval, bvec, b0_threshold=50):
+                save_metrics = ['S0_predicted', 'perfusion_fraction', 'D_star',
+                                'D']
+
+            S0_predicted = ivimfit.S0_predicted
```
Do you really need to refer to these here? You could just refer directly to the `ivimfit` attributes below as needed.

ShreyasFadnavis replied:
Will do as per your suggestion! Thanks :)

arokem commented on dipy/workflows/reconst.py:
```diff
@@ -861,3 +862,122 @@ def get_fitted_tensor(self, data, mask, bval, bvec, b0_threshold=50):
+            D = ivimfit.D
+
+            if 'S0_predicted' in save_metrics:
+                save_nifti(oS0_predicted, S0_predicted.astype(np.float32),
```
```suggestion
                save_nifti(oS0_predicted, ivimfit.S0_predicted.astype(np.float32),
```

arokem commented on dipy/workflows/reconst.py:
```diff
@@ -861,3 +862,122 @@ def get_fitted_tensor(self, data, mask, bval, bvec, b0_threshold=50):
+
+            if 'perfusion_fraction' in save_metrics:
+                save_nifti(operfusion_fraction,
+                           perfusion_fraction.astype(np.float32), affine)
```
```suggestion
                           ivimfit.perfusion_fraction.astype(np.float32), affine)
```

arokem commented on dipy/workflows/reconst.py:
```diff
@@ -861,3 +862,122 @@ def get_fitted_tensor(self, data, mask, bval, bvec, b0_threshold=50):
+                           perfusion_fraction.astype(np.float32), affine)
+
+            if 'D_star' in save_metrics:
+                save_nifti(oD_star, D_star.astype(np.float32), affine)
```
```suggestion
                save_nifti(oD_star, ivimfit.D_star.astype(np.float32), affine)
```

arokem commented on dipy/workflows/reconst.py:
```diff
@@ -861,3 +862,122 @@ def get_fitted_tensor(self, data, mask, bval, bvec, b0_threshold=50):
+                save_nifti(oD_star, D_star.astype(np.float32), affine)
+
+            if 'D' in save_metrics:
+                save_nifti(oD, D.astype(np.float32), affine)
```
```suggestion
                save_nifti(oD, ivimfit.D.astype(np.float32), affine)
```

ShreyasFadnavis commented:
@arokem Done! Thank you for all the help :+1:

skoudoro commented on dipy/workflows/tests/test_reconst_ivim.py:
```diff
@@ -0,0 +1,84 @@
+        np.savetxt(tmp_bval_path, bvals)
+        np.savetxt(tmp_bvec_path, bvecs.T)
+
+        ivim_flow._force_overwrite = True
```
Your test is great. I think the part between line [70](https://github.com/nipy/dipy/pull/1739/files#diff-621f294e8cafe159740bd93f1664cee6R70)-[80](https://github.com/nipy/dipy/pull/1739/files#diff-621f294e8cafe159740bd93f1664cee6R80) is useless so you can remove it.

ShreyasFadnavis replied:
Will do!

skoudoro commented on dipy/workflows/reconst.py:
```diff
@@ -861,3 +862,118 @@ def get_fitted_tensor(self, data, mask, bval, bvec, b0_threshold=50):
+        return 'ivim'
+
+    def run(self, input_files, bvalues_files, bvectors_files, split_b_D=400,
+            split_b_S0=200, save_metrics=[],
```
It would be good if you can add `mask_files` and `b0_threshold=50` as input arguments

ShreyasFadnavis replied:
done! :+1:

skoudoro commented on dipy/workflows/reconst.py:
```diff
@@ -861,3 +862,118 @@ def get_fitted_tensor(self, data, mask, bval, bvec, b0_threshold=50):
+            data, affine = load_nifti(dwi)
+
+            ivimfit, _ = self.get_fitted_ivim(data, bval, bvec,
+                                              b0_threshold=0)
```
Why `b0_threshold=0`?

ShreyasFadnavis replied:
> Why `b0_threshold=0`?

Done :+1:

skoudoro commented:
LGTM, Any other comments? I will wait until Wednesday to merge this PR.

ShreyasFadnavis commented:
@arokem @skoudoro @Garyfallidis I wanted some help with understanding why `40b1d7e` is failing only for Python 3.4?

Any feedback would be helpful!

arokem commented:
Hmm. I have no idea. Is this even in the code that you wrote?

I'll try to take another look tomorrow in a virtualenv with Python 3.4 installed.

arokem commented:
Seems like it was intermittent. I reran that job and Travis seems happy now. I'll go ahead and merge.

ShreyasFadnavis commented:
> Seems like it was intermittent. I reran that job and Travis seems happy now. I'll go ahead and merge.

Thank you so much 👍
